include/eos/fitting/ceres_nonlinear.hpp

```diff
@@ -270,51 +292,53 @@ struct ImageCost
-        const tmat4x4<T> rot_mtx = glm::mat4_cast(rot_quat);
+        const auto rot_mtx = glm::mat4_cast(rot_quat);
 
         // Todo: use get_opencv_viewport() from nonlin_cam_esti.hpp.
-        const tvec4<T> viewport(0, image.rows, image.cols, -image.rows); // OpenCV convention
+        const auto viewport = tvec4<T>(0, image.rows, image.cols, -image.rows); // OpenCV convention
 
         tvec3<T> projected_point;
         if (use_perspective)
         {
             const auto t_mtx = glm::translate(tvec3<T>(camera_translation_and_intrinsics[0],
                                                        camera_translation_and_intrinsics[1],
                                                        camera_translation_and_intrinsics[2]));
-            const T& focal = camera_translation_and_intrinsics[3];
-            const auto persp_mtx = glm::perspective(focal, T(aspect_ratio), T(0.1), T(1000.0));
+            const auto& focal = camera_translation_and_intrinsics[3];
+            const auto persp_mtx = glm::perspective(focal, static_cast<T>(aspect_ratio),
+                                                    static_cast<T>(0.1), static_cast<T>(1000.0));
             projected_point = glm::project(point_3d, t_mtx * rot_mtx, persp_mtx, viewport);
         } else
         {
-            const T& frustum_scale = camera_translation_and_intrinsics[2];
+            const auto& frustum_scale = camera_translation_and_intrinsics[2];
             const auto t_mtx = glm::translate(
                 tvec3<T>(camera_translation_and_intrinsics[0], camera_translation_and_intrinsics[1],
                          0.0)); // we don't have t_z in ortho camera, it doesn't matter where it is
-            const auto ortho_mtx =
-                glm::ortho(-1.0 * aspect_ratio * frustum_scale, 1.0 * aspect_ratio * frustum_scale,
-                           -1.0 * frustum_scale, 1.0 * frustum_scale);
+            const auto ortho_mtx = glm::ortho(-1.0 * aspect_ratio * frustum_scale, 1.0 * aspect_ratio * frustum_scale,
+                                              -1.0 * frustum_scale, 1.0 * frustum_scale);
             projected_point = glm::project(point_3d, t_mtx * rot_mtx, ortho_mtx, viewport);
         }
 
         // Access the image colour value at the projected pixel location, if inside the image - otherwise set
         // to (127, 127, 127) (maybe not ideal...):
-        if (projected_point.y < T(0) || projected_point.y >= T(image.rows) || projected_point.x < T(0) ||
-            projected_point.x >= T(image.cols))
+        if (projected_point.y < static_cast<T>(0) ||
+            projected_point.y >= static_cast<T>(image.rows) ||
+            projected_point.x < static_cast<T>(0) ||
+            projected_point.x >= static_cast<T>(image.cols))
         {
             // The point is outside the image.
-            residual[0] = T(127.0);
-            residual[1] = T(127.0);
-            residual[2] = T(127.0);
+            residual[0] = static_cast<T>(127.0);
```

Comment:
May I ask about the reason for the change here? I'd be quite curious to hear.

Reply:
As far as I remember it is a recommendation from "Efficient Modern C++" book. Just always use explicit casting because it is more noticeable and easier to find in the code.

It is not something critical :-) Can change it if you want.